### PR TITLE
feat: support output nvfp4 in trtllm-gen function call.

### PR DESCRIPTION
**Pull request by weireweire**
## 📌 Description
Only added fp4 support to trtllm_batch_decode_with_kv_cache, not added to wrapper yet.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [ ] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

### CODE REVIEW

**Comment by nvpohanh** [comment]
I will review tomorrow

File: flashinfer/decode.py [hunk 1]
```diff
@@ -1817,6 +1819,8 @@ def _paged_run(
             max_seq_len,
             bmm1_scale,
             bmm2_scale,
+            -1,
```
Comment (yzh119) [inline flashinfer/decode.py]:
```suggestion
            -1, # o_sf_scale
            -1, # o_sf_vec_size
```
Add some comments when arguments meaning is not clear.

File: csrc/trtllm_fmha_kernel_launcher.cu [hunk 2]
```diff
@@ -267,7 +289,7 @@ void trtllm_paged_attention_context(at::Tensor& out, at::Tensor& query, at::Tens
       o_data_type, TllmPagedAttentionMode::Context, batch_size, max_q_len, max_kv_len,
       num_pages_in_mem_pool, num_qo_heads, num_kv_heads, head_dim_qk, head_dim_vo, page_size,
       kv_stride_keys_values, kv_stride_heads, kv_stride_batch, max_num_blocks_per_seq, bmm1_scale,
-      bmm2_scale, window_left, sum_seq_q, sm_count, stream);
+      bmm2_scale, -1, -1, window_left, sum_seq_q, sm_count, stream);
```
Comment (yzh119) [inline csrc/trtllm_fmha_kernel_launcher.cu]:
```suggestion
      bmm2_scale, /* o_sf_scale =*/-1, /*o_sf_vec_size=*/-1, window_left, sum_seq_q, sm_count, stream);
```

File: tests/utils_fp4.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    scale = scale.to(torch.float8_e4m3fn).to(torch.float32)
+    output_scale = get_reciprocal(scale * get_reciprocal(global_scale))
+
+    scaled_x = sliced_x.to(torch.float32) * output_scale
+    clipped_x = torch.clamp(scaled_x, -6.0, 6.0).reshape(x.shape)
+    return cast_to_fp4(clipped_x), scale.squeeze(-1)
+
+
+def recover_swizzled_scales(scale, m, n, block_size):
+    round_up = lambda x, y: (x + y - 1) // y * y
```
Comment (yzh119) [inline tests/utils_fp4.py]: you can import it from `flashinfer.utils`

File: csrc/trtllm_fmha_kernel_launcher.cu [hunk 4]
```diff
@@ -236,7 +252,13 @@ void trtllm_paged_attention_context(at::Tensor& out, at::Tensor& query, at::Tens
   int num_qo_heads = query.size(1);
   int sum_seq_q = query.size(0);
   int head_dim_qk = query.size(2);
+  if (q_data_type == Data_type::DATA_TYPE_E2M1) {
+    head_dim_qk *= 2;  // assume even
```
Comment (yzh119) [inline csrc/trtllm_fmha_kernel_launcher.cu]: Better to add a equality check with key.size(-1), to avoid errors when q and k's data type is different (mixed precision),

File: csrc/trtllm_fmha_kernel_launcher.cu [hunk 5]
```diff
@@ -236,7 +252,13 @@ void trtllm_paged_attention_context(at::Tensor& out, at::Tensor& query, at::Tens
...
+  }
   int head_dim_vo = out.size(-1);
+  if (o_data_type == Data_type::DATA_TYPE_E2M1) {
+    head_dim_vo *= 2;
```
Comment (yzh119) [inline csrc/trtllm_fmha_kernel_launcher.cu]: ditto, add an equality check with value.size(-1)

File: csrc/trtllm_fmha_kernel_launcher.cu [hunk 6]
```diff
@@ -257,7 +279,7 @@ void trtllm_paged_attention_context(at::Tensor& out, at::Tensor& query, at::Tens
   const auto stream = at::cuda::getCurrentCUDAStream(device.index());
 
   trtllm_paged_attention_launcher(
-      out.data_ptr(), query.data_ptr(), key_value_cache.data_ptr(),
+      out.data_ptr(), nullptr, query.data_ptr(), key_value_cache.data_ptr(),
```
Comment (yzh119) [inline csrc/trtllm_fmha_kernel_launcher.cu]:
```suggestion
      out.data_ptr(), /*out_scale_factor=*/nullptr, query.data_ptr(), key_value_cache.data_ptr(),
```

File: csrc/trtllm_fmha_kernel_launcher.cu [hunk 7]
```diff
@@ -170,14 +175,19 @@ inline Data_type torch_dtype_to_tllm_data_type(at::ScalarType dtype) {
...
+    return Data_type::DATA_TYPE_E2M1;
   }
   return Data_type::DATA_TYPE_UNKNOWN;
 }
 
-void trtllm_paged_attention_decode(at::Tensor& out, at::Tensor& query, at::Tensor& key_value_cache,
-                                   at::Tensor& workspace_buffer, at::Tensor& block_tables,
-                                   at::Tensor& seq_lens, int64_t max_kv_len, double bmm1_scale,
-                                   double bmm2_scale, int64_t window_left, int64_t sm_count) {
+void trtllm_paged_attention_decode(at::Tensor& out, at::Tensor& out_scale_factor, at::Tensor& query,
```
Comment (yzh119) [inline csrc/trtllm_fmha_kernel_launcher.cu]: the type of `out_scale_factor` could be `std::optional<at::Tensor>` so that we can pass None instead of empty(0) at python side.

**Review by yzh119 — COMMENTED** [review]
Overall LGTM, @weireweire thanks for your contribution!
Left some suggestions for improvements.

**Comment by yzh119** [comment]
Also, there appear to be conflicts with the main branch in tests/test_fp4_quantize and tests/test_trtllm_gen_decode. Please rebase to resolve them.

File: csrc/trtllm_fmha_kernel_launcher.cu (hunk 7 above)
Comment (weireweire) [reply csrc/trtllm_fmha_kernel_launcher.cu]: then we cannot use & here, it's ok right?

Comment (weireweire) [reply csrc/trtllm_fmha_kernel_launcher.cu]: and maybe we should remove & for other tensor to keep consistency

File: csrc/trtllm_fmha_kernel_launcher.cu (hunk 4 above)
Comment (weireweire) [reply csrc/trtllm_fmha_kernel_launcher.cu]: we only have a key_value_cache as argument, I think we can only check with that and not really support k-dim != v-dim

File: csrc/trtllm_fmha_kernel_launcher.cu (hunk 5 above)
Comment (weireweire) [reply csrc/trtllm_fmha_kernel_launcher.cu]: ditto

File: tests/utils_fp4.py (hunk 3 above)
Comment (weireweire) [reply tests/utils_fp4.py]: note the filename is util_fp4.py, cause I want all util become util_xx.py so that the are in similar position in file explore. Please comment if you still want xx_util.py, or I can also rename those to util_xx.py

File: csrc/trtllm_fmha_kernel_launcher.cu (hunk 7 above)
Comment (nvpohanh) [reply csrc/trtllm_fmha_kernel_launcher.cu]: It seems that the convention is:

- Use `at::Tensor const&` for required non-mutable tensor.
- Use `at::Tensor&` for required mutable tensor.
- Use `at::optional<at::Tensor> const&` for optional non-mutable tensor.

What about optional mutable tensor? I have not seen an example...

See: https://github.com/flashinfer-ai/flashinfer/blob/main/csrc/trtllm_fused_moe_kernel_launcher.cu#L906-L907

File: csrc/trtllm_fmha_kernel_launcher.cu [hunk 8]
```diff
@@ -190,7 +200,13 @@ void trtllm_paged_attention_decode(at::Tensor& out, at::Tensor& query, at::Tenso
   int sum_seq_q = batch_size * q_len_per_request;
   int num_qo_heads = query.size(2);
   int head_dim_qk = query.size(3);
+  if (q_data_type == Data_type::DATA_TYPE_E2M1) {
```
Comment (nvpohanh) [inline csrc/trtllm_fmha_kernel_launcher.cu]: Add a comment explaining why we need to multiply by 2 here:

```
// Multiply by two to recover actual dims because FP4 tensor is stored as UINT8 dtype.
```

File: csrc/trtllm_fmha_kernel_launcher.cu (hunk 7 above)
Comment (weireweire) [reply csrc/trtllm_fmha_kernel_launcher.cu]: > optional mutable tensor

yes that's the problem, non-const ref cannot ref the right value that constructed when create std::optional.

And optional\<at::Tensor\> & won't avoid the tensor copy.

And adding const won't working as we always got raw pointer from tensor.

And at::Tensor is actually smart pointer, so we don't need & in fact. And I see other API is just use at::Tensor

@yzh119 please give use a standard for this.

File: tests/utils_fp4.py (hunk 3 above)
Comment (weireweire) [reply tests/utils_fp4.py]: also please comment if we need to make fp4_quantize op also return a FP4Tensor wrapper instead of two tensor

File: flashinfer/decode.py [hunk 9]
```diff
@@ -1971,17 +1975,55 @@ def trtllm_batch_decode_with_kv_cache(
     bmm2_scale: float,  # todo(Yingyi): add dynamic scale tensor later
     window_left: int = -1,
     out: Optional[torch.Tensor] = None,
-) -> torch.Tensor:
+    out_dtype: str = "auto",
```
Comment (nvpohanh) [inline flashinfer/decode.py]: Should we document what "auto" means? Something like:

```
# out_dtype must be one of ["auto", "fp4"].
# If out_dtype is "auto", output dtype is inferred from "out" tensor if it is provided, or inferred from
# "query" tensor if "out" is not provided.
# If out_dtype is "fp4", then returned dtype is assumed to be a FP4Tensor and the "out" tensor is
# assumed to be a FP4 tensor stored as UINT8 dtype.
```

File: flashinfer/decode.py [hunk 10]
```diff
@@ -1971,17 +1975,55 @@ def trtllm_batch_decode_with_kv_cache(
...
+    o_sf_vec_size: Optional[int] = None,
+) -> Union[torch.Tensor, FP4Tensor]:
     run_func = get_trtllm_gen_fmha_module().trtllm_paged_attention_decode
     sm_count = get_device_sm_count(query.device)
 
-    if out is None:
-        out = torch.empty_like(query)
-    else:
+    if out_dtype == "fp4":
+        assert query.dtype == torch.float8_e4m3fn
```
Comment (nvpohanh) [inline flashinfer/decode.py]:
```suggestion
        assert query.dtype == torch.float8_e4m3fn, "query must by FP8 when out_dtype is fp4."
```

File: flashinfer/decode.py [hunk 11]
```diff
@@ -1971,17 +1975,55 @@ def trtllm_batch_decode_with_kv_cache(
...
+        assert o_sf_scale is not None
+        assert o_sf_vec_size in [None, 16]
```
Comment (nvpohanh) [inline flashinfer/decode.py]:
```suggestion
        assert o_sf_vec_size in [None, 16], "only o_sf_vec_size = 16 is supported"
```

File: flashinfer/decode.py [hunk 12]
```diff
@@ -1990,10 +2032,16 @@ def trtllm_batch_decode_with_kv_cache(
...
+        o_sf_scale or -1.0,
+        o_sf_vec_size or -1,
         window_left,
         sm_count,
     )
-    return out
+
+    if out_dtype == "fp4":
+        return FP4Tensor(out, out_scale_factor, query.shape)
+    else:
```
Comment (nvpohanh) [inline flashinfer/decode.py]: this `else` is redundant:

```
if out_dtype == "fp4":
        return FP4Tensor(out, out_scale_factor, query.shape)

return out
```
or
```
return FP4Tensor(out, out_scale_factor, query.shape) if out_dtype == "fp4" else out
```

File: flashinfer/decode.py [hunk 13]
```diff
@@ -1971,17 +1975,55 @@ def trtllm_batch_decode_with_kv_cache(
     bmm2_scale: float,  # todo(Yingyi): add dynamic scale tensor later
     window_left: int = -1,
     out: Optional[torch.Tensor] = None,
```
Comment (nvpohanh) [inline flashinfer/decode.py]: just a thought: If we return `FP4Tensor`, should we also just ask user to provide `FP4Tensor` when output dtype is fp4? Otherwise, the asymmetry bugs me a little bit.

File: flashinfer/decode.py [hunk 14]
```diff
@@ -1971,17 +1975,55 @@ def trtllm_batch_decode_with_kv_cache(
...
+        # Use uint8 as the container dtype to compliant with next fp4 gemm.
+        _check_shape_dtype_device(
+            out_scale_factor,
+            fp4_out_scale_shape,
+            torch.float8_e4m3fn,
+            query.device,
+            "out_scale_factor",
+        )
+    elif out_dtype == "auto":
+        assert out_scale_factor is None
```
Comment (nvpohanh) [inline flashinfer/decode.py]: maybe to tight? if user already provides out/out_scale_factor, we should be able to infer out_dtype from it, right?

File: flashinfer/utils.py [hunk 15]
```diff
@@ -514,3 +514,55 @@ def pad_up(x: int, y: int) -> int:
 
 def get_device_sm_count(device: torch.device) -> int:
     return torch.cuda.get_device_properties(device).multi_processor_count
+
+
+class FP4Tensor:
```
Comment (nvpohanh) [inline flashinfer/utils.py:519]: Maybe call it NVFP4Tensor? Just to distinguish it from MXFP4 (which is E2M1 data + E8M0 sf) in case we add that later?

File: flashinfer/utils.py [hunk 16]
```diff
@@ -514,3 +514,55 @@ def pad_up(x: int, y: int) -> int:
...
+    """Wrapper class for FP4 tensors.
+
+    Since PyTorch doesn't natively support FP4, this wrapper contains:
+    - data: uint8 tensor storing the compressed FP4 data, the size of innermost dimension is ceil(original_dim / 2) since each uint8 stores 2 FP4 values
+    - scale: float8_e4m3fn tensor storing the scale factors
```
Comment (nvpohanh) [inline flashinfer/utils.py:524]: we should document the expected shape for this.

File: flashinfer/utils.py (hunk 15 above)
Comment (weireweire) [reply flashinfer/utils.py:519]: I think we can share this class and add a "type" to it, like torch tensor.

File: flashinfer/utils.py [hunk 17]
```diff
@@ -514,3 +514,55 @@ def pad_up(x: int, y: int) -> int:
...
+            if data.shape[-1] != expected_data_dim:
+                raise ValueError(
+                    f"data last dimension must be ceil(original_shape[-1] / 2). "
+                    f"data.shape[-1]={data.shape[-1]}, original_shape[-1]={original_shape[-1]}, "
+                    f"expected={expected_data_dim}"
+                )
+
+        self.data = data
+        self.scale = scale
+        self.original_shape = original_shape
```
Comment (nvpohanh) [inline flashinfer/utils.py:568]: should we add a `shape` property, like what TRT-LLM did? so that user can treat it as a normal torch tensor and do: `tensor.shape`?

See: https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/_torch/utils.py#L97

File: flashinfer/decode.py (hunk 13 above)
Comment (weireweire) [reply flashinfer/decode.py]: I think we can do that and remove out_scale_factor arg

File: flashinfer/utils.py (hunk 16 above)
Comment (weireweire) [reply flashinfer/utils.py:524]: but that's depend on the vec_size and quant method, so currently we check in the decode API.

File: flashinfer/utils.py (hunk 17 above)
Comment (weireweire) [reply flashinfer/utils.py:568]: The reason I didn't add it here is I'm not sure if we could assume the original_shape is even at last dim.
The logic could be:
```
if self.original_shape: return self.original_shape
else return self.data.shape[:-1] + self.data.shape[-1] * 2
```
but if the shape is not even, the *2 here will be wrong, so I won't do that infer so that user will notice the difference.

what trtllm did is simply return self.data.shape, but that's the uint8 shape which is error prone

**Comment by weireweire** [comment]
@nvpohanh @yzh119 I added a new commit to optimize the trtllm_batch_decode_with_kv_cache API to use different type, please have a look.

File: flashinfer/utils.py (hunk 15 above)
Comment (weireweire) [reply flashinfer/utils.py:519]: added dtype = "nvfp4"

File: flashinfer/decode.py (hunk 9 above)
Comment (weireweire) [reply flashinfer/decode.py]: added doc

File: csrc/trtllm_fmha_kernel_launcher.cu (hunk 7 above)
Comment (yzh119) [reply csrc/trtllm_fmha_kernel_launcher.cu]: Just use `at::Tensor` is fine, we directly use at::Tensor for most of the bindings codes, no need to use reference here as you mentioned at::Tensor is a smart pointer..

File: tests/utils_fp4.py (hunk 3 above)
Comment (yzh119) [reply tests/utils_fp4.py]: > also please comment if we need to make fp4_quantize op also return a FP4Tensor wrapper instead of two tensor

I welcome such design, but please make a note that we have breaking changes on the interface to let user know about it.

Comment (yzh119) [reply tests/utils_fp4.py]: > note the filename is util_fp4.py, cause I want all util become util_xx.py so that the are in similar position in file explore. Please comment if you still want xx_util.py, or I can also rename those to util_xx.py

I have no preference so far. utils_fp4.py sounds good to me.

**Comment by weireweire** [comment]
@nvpohanh @yzh119 if no more issue, please approve and merge this

File: flashinfer/utils.py (hunk 17 above)
Comment (nvpohanh) [reply flashinfer/utils.py:568]: agreed.

**Comment by nvpohanh** [comment]
LGTM! thanks @yzh119 to do final review

**Comment by yyihuang** [comment]
Fixed error by #1348
